icicle/src/poseidon2/poseidon.cu

```diff
@@ -0,0 +1,76 @@
+#include "poseidon2/poseidon2.cuh"
+#include "constants.cu"
+#include "kernels.cu"
+
+namespace poseidon2 {
+  static int poseidon_block_size = 128;
+
+  template <typename S, int T>
+  int poseidon_number_of_blocks(size_t number_of_states)
+  {
+    return number_of_states / poseidon_block_size + static_cast<bool>(number_of_states % poseidon_block_size);
+  }
+
+  template <typename S, int T>
+  cudaError_t permute_many(
+    S* states, S* states_out, size_t number_of_states, const Poseidon2Constants<S>& constants, cudaStream_t& stream)
+  {
+    poseidon2_permutation_kernel<S, T>
+      <<<poseidon_number_of_blocks<S, T>(number_of_states), poseidon_block_size, 0, stream>>>(
+        states, states_out, number_of_states, constants);
+    CHK_IF_RETURN(cudaPeekAtLastError());
+    return CHK_LAST();
+  }
+
+  template <typename S, int T>
+  cudaError_t poseidon2_hash(
+    S* states,
```

Comment:
better be const S* if states is not modified

Reply:
states can be modified